// @shopify/hydrogen/dist/esnext/components/CartProvider/hooks.client.js
import React, { useState } from 'react';
import { useShop } from '../../foundation';
import { flattenConnection } from '../../utilities';
import { CartCreate, defaultCartFragment } from './cart-queries';
import { SHOPIFY_STOREFRONT_ID_HEADER, STOREFRONT_API_PUBLIC_TOKEN_HEADER, } from '../../constants';
export function useCartFetch() {
    const { storeDomain, storefrontApiVersion, storefrontToken, storefrontId } = useShop();
    return React.useCallback(({ query, variables, }) => {
        const headers = {
            'Content-Type': 'application/json',
            'X-SDK-Variant': 'hydrogen',
            'X-SDK-Version': storefrontApiVersion,
            [STOREFRONT_API_PUBLIC_TOKEN_HEADER]: storefrontToken,
        };
        if (storefrontId) {
            headers[SHOPIFY_STOREFRONT_ID_HEADER] = storefrontId;
        }
        return fetch(`https://${storeDomain}/api/${storefrontApiVersion}/graphql.json`, {
            method: 'POST',
            headers,
            body: JSON.stringify({
                query: query.toString(),
                variables,
            }),
        })
            .then((res) => res.json())
            .catch((error) => {
            return {
                data: undefined,
                error: error.toString(),
            };
        });
    }, [storeDomain, storefrontApiVersion, storefrontToken, storefrontId]);
}
export function useInstantCheckout() {
    const [cart, updateCart] = useState();
    const [checkoutUrl, updateCheckoutUrl] = useState();
    const [error, updateError] = useState();
    const fetch = useCartFetch();
    const createInstantCheckout = React.useCallback(async (cartInput) => {
        const { data, errors } = await fetch({
            query: CartCreate(defaultCartFragment),
            variables: {
                input: cartInput,
            },
        });
        if (errors) {
            updateError(errors);
            updateCart(undefined);
            updateCheckoutUrl(undefined);
        }
        if (data?.cartCreate?.cart) {
            const dataCart = data.cartCreate.cart;
            updateCart({
                ...dataCart,
                // @ts-expect-error While the cart still uses fragments, there will be a TS error here until we remove those fragments and get the type in-line
                lines: flattenConnection(dataCart.lines),
                note: dataCart.note ?? undefined,
            });
            updateCheckoutUrl(dataCart.checkoutUrl);
        }
    }, [fetch]);
    return { cart, checkoutUrl, error, createInstantCheckout };
}
