// @shopify/cli-kit/dist/index.js
export { b as abort, d as analytics, e as api, f as archiver, g as checksum, h as cli, c as constants, i as dependency, j as dotenv, k as environment, l as error, m as file, n as git, o as github, p as haiku, q as http, r as id, s as npm, t as os, u as output, v as path, w as plugins, x as port, y as ruby, z as schema, A as semver, B as session, C as store, D as string, E as system, G as template, H as temporary, I as toml, J as ui, K as version, L as vscode, M as yaml } from './index-eb280c22.js';
import * as Fastify from 'fastify';
export { Fastify as fastify };
import 'path';
import 'url';
import 'node:path';
import 'node:url';
import 'node:process';
import 'node:fs';
import 'os';
import 'util';
import 'stream';
import 'fs';
import 'events';
import 'node:os';
import '@oclif/core';
import 'source-map-support';
import 'constants';
import 'assert';
import 'node:stream';
import 'node:util';
import 'crypto';
import 'prettier';
import 'tty';
import 'stacktracey';
import 'node:buffer';
import 'node:child_process';
import 'child_process';
import 'buffer';
import 'node:http';
import 'node:https';
import 'node:zlib';
import 'node:net';
import 'http';
import 'https';
import 'readline';
import 'inquirer';
import 'dns';
import 'zlib';
import 'http2';
import 'tls';
import 'net';
import 'punycode';
import 'envfile';
import 'module';
