// @shopify/cli-hydrogen/dist/commands/hydrogen/preview.js
import { error, path, file, output, system, cli } from '@shopify/cli-kit';
import { fileURLToPath } from 'url';
import { h as hydrogenFlags } from '../../flags-39d1e3c2.js';
import { Flags, Command } from '@oclif/core';

async function previewInNode({ directory, port }) {
  const buildOutputPath = await path.resolve(directory, "dist/node");
  if (!await file.exists(buildOutputPath)) {
    output.info(output.content`Couldn’t find a Node.js server build for this project. Running ${output.token.packagejsonScript("yarn", "shopify hydrogen build", "--target=node")} to create one.`);
    await system.exec("yarn", ["shopify", "hydrogen", "build", "--target=node"], {
      cwd: directory,
      stdout: process.stdout
    });
  }
  await system.exec("node", ["--enable-source-maps", buildOutputPath], {
    env: { PORT: `${port}` },
    cwd: directory,
    stdout: process.stdout
  });
}
async function previewInWorker({ directory, port }) {
  const config = {
    port,
    workerFile: "dist/worker/index.js",
    assetsDir: "dist/client",
    buildCommand: "yarn build",
    modules: true,
    watch: true,
    buildWatchPaths: ["./src"],
    autoReload: true
  };
  await file.write(path.resolve(directory, "mini-oxygen.config.json"), JSON.stringify(config, null, 2));
  function cleanUp(options) {
    if (options.exit) {
      file.remove(path.resolve(directory, "mini-oxygen.config.json"));
    }
  }
  process.on("SIGINT", cleanUp.bind(null, { exit: true }));
  const executable = await oxygenPreviewExecutable();
  await system.exec(executable, [], {
    env: { NODE_OPTIONS: "--experimental-vm-modules" },
    cwd: directory,
    stdout: process.stdout
  });
}
const OxygenPreviewExecutableNotFound = new error.Bug("Could not locate the executable file to run Oxygen locally.");
async function oxygenPreviewExecutable() {
  const cwd = path.dirname(fileURLToPath(import.meta.url));
  const executablePath = await path.findUp("node_modules/.bin/oxygen-preview", { type: "file", cwd });
  if (!executablePath) {
    throw OxygenPreviewExecutableNotFound;
  }
  return executablePath;
}

const _Preview = class extends Command {
  async run() {
    const { flags } = await this.parse(_Preview);
    const directory = flags.path ? path.resolve(flags.path) : process.cwd();
    const port = parseInt(flags.port, 10);
    if (flags.target === "worker") {
      await previewInWorker({ directory, port });
    } else if (flags.target === "node") {
      await previewInNode({ directory, port });
    }
  }
};
let Preview = _Preview;
Preview.description = "Run a Hydrogen storefront locally in a worker environment";
Preview.flags = {
  ...cli.globalFlags,
  path: hydrogenFlags.path,
  port: Flags.string({
    char: "p",
    hidden: true,
    description: "the port to run the preview server on",
    default: "3000",
    env: "SHOPIFY_FLAG_PORT"
  }),
  target: Flags.string({
    char: "t",
    description: "the target environment (worker or node)",
    options: ["node", "worker"],
    default: "worker",
    env: "SHOPIFY_FLAG_PREVIEW_TARGET"
  })
};

export { Preview as default };
